### ISSUE TITLE
aws s3 BackupStorageLocation behind proxy: Unable to determine bucket's region

### ISSUE THREAD

**Issue by aayushrangwala** [vmware-tanzu/velero#3449]
**What steps did you take and what happened:**
[A clear and concise description of what the bug is, and what commands you ran.)

While running velero from the official [helm chart](https://github.com/vmware-tanzu/helm-charts/tree/main/charts/velero), behind a local (squid) proxy on mac using a self-signed certificate. As per the [docs](https://velero.io/docs/v1.5/self-signed-certificates/#docs) for passing cacert to velero, I am adding it to the BackupStorageLocation, but we are getting continuous errors as below:

```
time="2021-02-11T11:17:36Z" level=error msg="Error getting backup store for this location" backupLocation=dp-unstable-dev controller=backup-sync error="rpc error: code = Unknown desc = unable to determine bucket's region" error.file="/go/src/github.com/vmware-tanzu/velero-plugin-for-aws/velero-plugin-for-aws/helpers.go:52" error.function=main.GetBucketRegion logSource="pkg/controller/backup_sync_controller.go:167"
```

This issue is a part of another issue which was solved ref: #3433

It is found in the code of the plugin that, while determining region, it create an aws session which donot consider the ca certs passed in the BSL or any configuration.
Since the certificates which are needed to be passed for the s3 region api to work are known but its not added in the session config.

**What did you expect to happen:**

velero should be able to get the correct region for the bucket and prefix passed in the BSL

**The output of the following commands will help us better understand what's going on**:
(Pasting long output into a [GitHub gist](https://gist.github.com) or other pastebin is fine.)

- `kubectl logs deployment/velero -n velero`: [Here](https://gist.github.com/aayushrangwala/1307ac07dab8397913946a1c9719e36e) are the logs

**Environment:**

  - Velero version: v1.5.3 both client and server
  - Velero features: <NOT SET>
  -  Kubernetes version: kubectl version: v1.20 and server: v1.19
  - Kubernetes installer & version: Kind: kind v0.9.0 go1.15.2 darwin/amd64
  - Cloud provider or hardware configuration: Mac/OSX

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "I would like to see this bug fixed as soon as possible"
- :-1: for "There are more important bugs to focus on right now"

**Comment by zubron** [comment]
@aayushrangwala - Is there a reason why you are relying on the plugin determining the region and can't add the region to the BSL config as described in the AWS plugin docs? https://github.com/vmware-tanzu/velero-plugin-for-aws#install-and-start-velero

Looking at where this error is generated in the AWS plugin, it should only attempt to find the region if both the `s3Url` and `region` are not specified in config: https://github.com/vmware-tanzu/velero-plugin-for-aws/blob/main/velero-plugin-for-aws/object_store.go#L128-L137
By providing either of those, this check should be skipped.

**Comment by aayushrangwala** [comment]
> @aayushrangwala - Is there a reason why you are relying on the plugin determining the region and can't add the region to the BSL config as described in the AWS plugin docs? https://github.com/vmware-tanzu/velero-plugin-for-aws#install-and-start-velero
> 
> Looking at where this error is generated in the AWS plugin, it should only attempt to find the region if both the `s3Url` and `region` are not specified in config: https://github.com/vmware-tanzu/velero-plugin-for-aws/blob/main/velero-plugin-for-aws/object_store.go#L128-L137
> By providing either of those, this check should be skipped. This should unblock you in the meantime.

@zubron  Thanks for the response. Yes, I checked that code and got unblocked for now, but raised this bug as we are working on a feature where the user might not always give the region and the cluster workloads must be running behind the proxy. Which makes us blocked by this issue as, if the aws session created in the plugin doesn't use the certs (as discussed in the referred ticket), the BSL wont be in Available state.

So is this something which needs a bit of discussion and more suggestions before even qualifying it as a bug or it can be considered and he actual work is started on it?

**Comment by nrb** [comment]
@aayushrangwala Thanks for getting the PR started!

I think this might be useful for BSLs behind a proxy, but I'm concerned about the volumesnapshot calls (if any) also being affected. I think if we were to adopt this, then we'd want to make sure _all_ AWS calls were supported.

Does your use case only use S3, without EBS volumes?

**Comment by aayushrangwala** [comment]
> @aayushrangwala Thanks for getting the PR started!
> 
> I think this might be useful for BSLs behind a proxy, but I'm concerned about the volumesnapshot calls (if any) also being affected. I think if we were to adopt this, then we'd want to make sure _all_ AWS calls were supported.
> 
> Does your use case only use S3, without EBS volumes?

@nrb I think we should support the VolumeSnapshot API calls also along with the BSL ones. Because if we are adopting the certs support it should be more generic rather than a specific use case. I will add one more commit adding the support for cacerts in volumeSnapshot calls too

**Comment by zubron** [comment]
@aayushrangwala With regard to the VolumeSnapshot APIs, from looking at the AWS SDK docs, it's not clear whether there is a way to detect the region, and the region is a [required parameter when creating the config for a session](https://github.com/aws/aws-sdk-go/blob/v1.37.17/aws/config.go#L63-L70). We can avoid specifying the region with S3 related calls as the SDK provides the [`GetBucketRegion`](https://pkg.go.dev/github.com/aws/aws-sdk-go/service/s3/s3manager#GetBucketRegion) function but there may not be a similar approach for interacting with snapshots.

**Comment by aayushrangwala** [comment]
> @aayushrangwala With regard to the VolumeSnapshot APIs, from looking at the AWS SDK docs, it's not clear whether there is a way to detect the region, and the region is a [required parameter when creating the config for a session](https://github.com/aws/aws-sdk-go/blob/v1.37.17/aws/config.go#L63-L70). We can avoid specifying the region with S3 related calls as the SDK provides the [`GetBucketRegion`](https://pkg.go.dev/github.com/aws/aws-sdk-go/service/s3/s3manager#GetBucketRegion) function but there may not be a similar approach for interacting with snapshots.

@zubron Yes I agree, we can assume that region will always be present in the VolumeSnapshot CR but still volumeSnapshotter needs to communicate outside the proxy, calling aws calls to provision snapshots, for that the aws session used, needs to have the certs. This problem was solved for us in BSL because it already has the caCert field and consumes it

**Comment by zubron** [comment]
@aayushrangwala Okay, that's good to know. I mentioned it due to your [previous comment](https://github.com/vmware-tanzu/velero/issues/3449#issuecomment-777648677) about the region not always being provided.

**Comment by aayushrangwala** [comment]
> @aayushrangwala Okay, that's good to know. I mentioned it due to your [previous comment](https://github.com/vmware-tanzu/velero/issues/3449#issuecomment-777648677) about the region not always being provided.

@zubron To add the same support in VolumeSnapshot, I believe, we need to add a new similar filed of caCert in VolumeSnapshot CRD. The cert file passed to velero install command will propagate the cert data to both the objects, BSL and VSL. And the aws plugin VolSnapshotter implementation will add the certs to the aws sessions.

Please suggest the next steps on the same.

**Comment by nrb** [comment]
> Yes, I checked that code and got unblocked for now, but raised this bug as we are working on a feature where the user might not always give the region and the cluster workloads must be running behind the proxy. 

I think this should be solved via validation - the region is a required parameter for the call, so we should communicate this to the user.

For volume snapshots, you are correct that the new field would need to be added. This is achievable, certainly, and I don't think the technical challenges for accepting and populating being too difficult.

The complication I see is using an MitM proxy that may cause issues; I am not currently aware of proxy support on AWS EBS calls, but we can research that. I think your use case doesn't require this right now, but it's a scenario we need to consider to get this parity.

Another complication to consider is node-based authentication, such as AWS's IAM AssumeRole support. How does this cert work interact with such a setup?

**Comment by reasonerjt** [comment]
I think the comment around volume snapshotter is valid, but I have a PR to solve the problem in OP and it's more consistent to respect the settings in BSL for GetBucket.

I'll close this issue when the PR is merged.
